green_cli/authenticators/wally.py

```diff
@@ -0,0 +1,172 @@
+import wallycore as wally
+
+from green_cli.authenticators import *
+
+class WallyAuthenticator(MnemonicOnDisk, HardwareDevice):
+    """Stores mnemonic on disk but does not pass it to the gdk
+
+    This class illustrates how the hardware device interface to the gdk can be used to implement all
+    required crypto operations external to the gdk and thus avoid passing any key material to the
+    gdk at all.
+    """
+
+    @property
+    def name(self):
+        return 'libwally software signer'
+
+    def create(self, session_obj):
+        """Create and register a new wallet"""
+        logging.warning("Generating mnemonic using gdk")
+        self._mnemonic = gdk.generate_mnemonic()
+        return self.register(session_obj)
+
+    @property
+    def seed(self):
+        _, seed = wally.bip39_mnemonic_to_seed512(self._mnemonic, None)
+        return seed
+
+    @property
+    def master_key(self):
+        return wally.bip32_key_from_seed(self.seed, wally.BIP32_VER_TEST_PRIVATE,
+                                         wally.BIP32_FLAG_KEY_PRIVATE)
+
+    def derive_key(self, path: List[int]):
+        if not path:
+            return self.master_key
+        else:
+            return wally.bip32_key_from_parent_path(self.master_key, path,
+                                                    wally.BIP32_FLAG_KEY_PRIVATE)
+
+    def get_xpub(self, path: List[int]):
+        return wally.bip32_key_to_base58(self.derive_key(path), wally.BIP32_FLAG_KEY_PUBLIC)
+
+    def get_privkey(self, path: List[int]) -> bytearray:
+        return wally.bip32_key_get_priv_key(self.derive_key(path))
+
+    def sign_message(self, path: List[int], message: str) -> bytearray:
+        message = message.encode('utf-8')
+        formatted = wally.format_bitcoin_message(message, wally.BITCOIN_MESSAGE_FLAG_HASH)
+        privkey = self.get_privkey(path)
+        signature = wally.ec_sig_from_bytes(privkey, formatted,
+                                            wally.EC_FLAG_ECDSA | wally.EC_FLAG_GRIND_R)
+        return wally.ec_sig_to_der(signature)
+
+    def _get_sighash(self, wally_tx, index, utxo):
+        flags = wally.WALLY_TX_FLAG_USE_WITNESS
+        prevout_script = wally.hex_to_bytes(utxo['prevout_script'])
+        return wally.tx_get_btc_signature_hash(
+                wally_tx, index, prevout_script, utxo['satoshi'], wally.WALLY_SIGHASH_ALL, flags)
+
+    def _sign_tx(self, details, wally_tx):
+        txdetails = details['transaction']
+        utxos = txdetails['used_utxos'] or txdetails['old_used_utxos']
```

Comment:
this should be `old_used_utxos + used_utxos` - see get_ga_signing_inputs() in gdk.